library/core/src/option.rs

```diff
@@ -689,6 +689,92 @@ impl<T> Option<T> {
         }
     }
 
+    const fn get_some_offset() -> isize {
+        if core::mem::size_of::<Option<T>>() == core::mem::size_of::<T>() {
+            // niche optimization means the `T` is always stored at the same position as the Option.
+            0
+        } else {
+            assert!(
+                core::mem::size_of::<Option<T>>()
+                    == core::mem::size_of::<Option<core::mem::MaybeUninit<T>>>()
+            );
+            let some_uninit = Some(core::mem::MaybeUninit::<T>::uninit());
+            // SAFETY: This gets the byte offset of the `Some(_)` value following the fact that
+            // niche optimization is not active, and thus Option<T> and Option<MaybeUninit<t>> share
+            // the same layout, which we have asserted above.
+            unsafe {
+                (some_uninit.as_ref().unwrap() as *const core::mem::MaybeUninit<T> as *const u8)
+                    .offset_from(
+                        &some_uninit as *const Option<core::mem::MaybeUninit<T>> as *const u8,
+                    )
```

Comment:
ymmv: thanks to #103489, the types to `byte_offset_from` don't need to match, so you might be able to remove some casts by phrasing this something like
```rust
                (some_uninit.as_ref().unwrap() as *const core::mem::MaybeUninit<T>)
                    .byte_offset_from(ptr::addr_of!(some_uninit))
```